### PR TITLE
Google login

### PR DESCRIPTION
**Pull request by waymao**
# Auth system with google login
Google login now seems to be working. 
Raw username and password login require a little more effort because of the hashing of the password. However, since all API and procedure will be the same as long as the user is logged in, I'm sending this pull request so we can get started on the other parts of our project where log in is required. 
Things have not been extensively tested yet. The errors below from Travis CI seems to come from the database branch side. This branch includes unmerged code from the database branch until 4PM Friday Afternoon.

## For backend/database folks:
If you have some ejs function like this:
```
router.get('/login', (req,res) => {
  res.render('login');
});
```
Then the user info will be available at `req.user`.
Database accessing functions and configs are at `config/passport-setup.js`. Other functions are at `routes/auth-routes.js`.

## For frontend/React folks:
After a successful login, the login page will receive a message from the pop-up window, and be redirected to the home page.
The API address to see whether the user is logged in: `http://localhost:5000/auth/loginstatus`.
If the user is not logged in, the response will be
```
{
  "status": "logged_out"
}
```
If the user is logged in, the response will be 
```
{
  "status": "logged_in", 
  "user": {
     /* user info */
  }
}
```
User info contains name, username and email address.

### CODE REVIEW

File: client/src/Authentication/LoginForm.js [hunk 1]
```diff
@@ -95,14 +94,28 @@ class LoginForm extends React.Component {
...
             </div>
         )
+    };
+
+    openGoogleLogin = () => {
+        window.open("http://localhost:5000/auth/google/", "Login",'height=800,width=500');
+    };
+
+    componentDidMount = () => {
+        window.addEventListener('message', (event) => {
```
Comment (StevieSong) [inline client/src/Authentication/LoginForm.js]: don't like this post back to the parent window, maybe consider a solution like this? https://stackoverflow.com/questions/28392393/passport-js-after-authentication-in-popup-window-close-it-and-redirect-the-pa

File: server/config/passport-setup.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    findUserOrCreate(profile.emails[0].value, profile, done);
+}
+
+function findUser(username)
+{
+    console.log(username);
+    return db.User.findByPk(username);
+}
+
+function findUserOrCreate (username, profile, done)
```
Comment (StevieSong) [inline server/config/passport-setup.js]: is this default behavior what we want? if user not found, auto create? @annelau99

File: server/index.js [hunk 3]
```diff
@@ -1,71 +1,56 @@
...
+const passportSetup = require('./config/passport-setup');
+const userRoutes = require('./routes/userRoutes');
+const groupRoutes = require('./routes/groupRoutes');
 let db;
 
+const passport = require('passport');
+const authRoutes = require('./routes/auth-routes');
+const session = require('express-session');
+
+app.set('view engine', 'ejs');
```
Comment (StevieSong) [inline server/index.js]: do we need a separate view engine?

File: server/routes/auth-routes.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  res.render('signup');
+});
+
+// auth with google
+router.get('/google', passport.authenticate('google', {
+  scope: ['profile', 'email']
+}));
+
+// callback route for google to redirect to
+router.get('/google/redirect',
```
Comment (StevieSong) [inline server/routes/auth-routes.js]: passport.authenticate() does the initial login, handle subsequent authentication by checking req.isAuthenticated(). This way, in building out the other api routes, we can use one authentication checker. see https://github.com/JumboCode/E4P/blob/2c7c7a5129d011bb599460ab7c2fffba09065fb4/routes/adminRoutes.js#L14

File: client/src/Authentication/LoginForm.js (hunk 1 above)
Comment (waymao) [reply client/src/Authentication/LoginForm.js]: I did try the method in the link. Right now the node.js server and React server are on different ports so I call to the main window's function will be blocked. The current solution is a workaround. In our final product, will the node server and the react server on the same domain name (via proxy or sth)?

File: server/index.js (hunk 3 above)
Comment (waymao) [reply server/index.js]: @harshyyy21 added this line

File: client/src/Authentication/LoginForm.js (hunk 1 above)
Comment (StevieSong) [reply client/src/Authentication/LoginForm.js]: ah interesting, in the end yes, the express backend will actually serve a static react app. the current separation is for development (i.e. the dev tool create-react-app runs a separate dev server) but in production, we'll build and deploy react app via express. see this [article](https://daveceddia.com/deploy-react-express-app-heroku/)

File: server/routes/auth-routes.js (hunk 4 above)
Comment (waymao) [reply server/routes/auth-routes.js]: Yes, I will be using this method in other rules. For google login, the authenticate has two phases, 1. redirecting to google, and 2. receive info (callback URL) from google. I believe both phases require passport.authenticate(), as is in the official docs/examples. `http://www.passportjs.org/docs/google/`

**Comment by waymao** [comment]
New changes made, will create a new pull request.

**Pull request closed without merging**